fixuid does not handle errors from filepath.Walk

Hi!  First, thanks for creating this eminently useful utility -- container bind mounts can be such a pain, and anything that makes dealing with mounts easier is most, most welcome.

I am hitting a fatal error when using `fixuid` as the entrypoint for a container that's based on the [ruby:2.4 image](https://github.com/docker-library/ruby/blob/1d0fe6a9d5c75d8a114fb03016f528d68c613d80/2.4/jessie/Dockerfile).  Here's an overview of my environment:

```shell
$ docker version
Client:
 Version:       17.12.0-ce
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    c97c6d6
 Built: Wed Dec 27 20:10:14 2017
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.0-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.2
  Git commit:   c97c6d6
  Built:        Wed Dec 27 20:12:46 2017
  OS/Arch:      linux/amd64
  Experimental: false
$ uname -a
Linux mydomain.com 3.10.0-693.11.1.el7.x86_64 #1 SMP Mon Dec 4 23:52:40 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

```

And here are the two Dockerfiles involved in the build:

```dockerfile
FROM debian:jessie                                                                                                                         
                                                                                                                                           
ARG FIXUID_VERSION                                                                                                                         
ENV FIXUID_VERSION 0.2                                                                                                                     
                                                                                                                                           
ADD "https://github.com/boxboat/fixuid/releases/download/v${FIXUID_VERSION}/fixuid-${FIXUID_VERSION}-linux-amd64.tar.gz" /tmp/fixuid.tar.gz
                                                                                                                                           
RUN     tar -C /usr/local/bin -xvf /tmp/fixuid.tar.gz \                                                                                    
    &&  rm -f /tmp/fixuid.tar.gz \                                                                                                         
    &&  chmod u+sx /usr/local/bin/fixuid                                                                                                   
                                                                                                                                           
ENTRYPOINT ["/usr/local/bin/fixuid"]                                                                                                       
```

```dockerfile                                                        
FROM custom/fixuid:custom AS fixuid                           
                                                              
FROM ruby:2.4                                                 
                                                                  
COPY --from=fixuid /usr/local/bin/fixuid /usr/local/bin/fixuid

# More stuff below
```

When I try to run a container based off of an image built from the second Dockerfile, I get a stack trace:

```shell
$ docker run --rm -it -u 1002:100 custom/ruby:2.4 bash
fixuid: fixuid should only ever be used on development systems. DO NOT USE IN PRODUCT
fixuid: updating user 'auser' to UID '1002                                         
fixuid: runtime GID '100' matches existing group 'agroup'; not changing GID           
panic: runtime error: invalid memory address or nil pointer dereference              
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x522b30]              
                                                                                     
goroutine 1 [running]:                                                               
main.main.func1(0xc4201074c0, 0x18, 0x0, 0x0, 0x613600, 0xc420080ba0, 0x0, 0x0)      
        /home/caleb/go/src/github.com/boxboat/fixuid/fixuid.go:185 +0x60             
path/filepath.walk(0xc420102515, 0x9, 0x614a80, 0xc42011c750, 0xc420016ba0, 0x0, 0x0)
        /usr/local/go/src/path/filepath/path.go:377 +0x2eb                           
path/filepath.walk(0xc420015850, 0x4, 0x614a80, 0xc4200f2750, 0xc420016ba0, 0x0, 0x0)
        /usr/local/go/src/path/filepath/path.go:381 +0x39a                           
path/filepath.walk(0x569e59, 0x1, 0x614a80, 0xc4200828f0, 0xc420016ba0, 0x0, 0x30)   
        /usr/local/go/src/path/filepath/path.go:381 +0x39a                           
path/filepath.Walk(0x569e59, 0x1, 0xc420016ba0, 0xc420082820, 0x0)                   
        /usr/local/go/src/path/filepath/path.go:403 +0x11d                           
main.main()                                                                          
        /home/caleb/go/src/github.com/boxboat/fixuid/fixuid.go:215 +0x1349           
```

This seems to result from the fact that `fixuid` does not (explicitly) handle errors passed to the `filepath.WalkFunc` by `filepath.Walk`.  I compiled a patched version of `fixuid` with the following change:

```go
// this function is called for every file visited                      
visit := func(filePath string, fileInfo os.FileInfo, err error) error {
    // bail out on walk error
    if err != nil {                                                    
        logger.Fatalln(err)                                            
    }
```

When I rebuild the two Docker images using this patched `fixuid`, I instead get the following error:

```shell
docker run --rm -it -u 1002:100 custom/ruby:2.4 bash
fixuid: fixuid should only ever be used on development systems. DO NOT USE IN PRODUCTION
fixuid: updating user 'auser' to UID '1002
fixuid: updating group 'agroup' to GID '1010
fixuid: lstat /etc/dpkg/shlibs.default: no such file or directory
```

And, indeed:

```shell
docker run --entrypoint '' --rm -it -u 1002:100 custom/ruby:2.4 ls -l /etc/dpkg
ls: cannot access /etc/dpkg/shlibs.default: No such file or directory
ls: cannot access /etc/dpkg/shlibs.override: No such file or directory
total 4
-rw-r--r-- 1 root root 446 Nov 28  2014 dpkg.cfg
drwxr-xr-x 2 root root  31 Dec 10 00:00 dpkg.cfg.d
drwxr-xr-x 2 root root  33 Dec 10 00:00 origins
?????????? ? ?    ?      ?            ? shlibs.default
?????????? ? ?    ?      ?            ? shlibs.override
```
I'm not quite sure what the deal is with the `ruby:2.4` image (`shlibs.default` and `shlibs.override` have the same weird `ls` output in both `custom/ruby:2.4` and the upstream `ruby:2.4`), but in any case it would be great if `fixuid` could handle errors like missing files that nonetheless "show up" (?) when walking the filesystem hierarchy.  I'm not a Gopher, but I'd be happy to submit a PR if I can get an idea of what `fixuid` should do (log and bail out? blithely ignore?) in scenarios like the one I've described.

Thanks very much!

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,7 +2,7 @@ language: go
 sudo: required
 dist: trusty
 go:
-  - 1.8
+  - 1.9
 
 branches:
   only:
@@ -15,6 +15,7 @@ install:
   - go get github.com/go-ozzo/ozzo-config
 
 script:
+  - docker info
   - ./test.sh
 
 notifications:

--- a/README.md
+++ b/README.md
@@ -94,6 +94,28 @@ nginx:
     - ./www:/var/www
 ```
 
+## Specify Paths and Behavior across Devices
+
+The default behavior of `fixuid` is to start at the root path `/` and recursively scan each file and directory on the same devices as `/`.  In the configuration file `/etc/fixuid/config.yml`, you can specify specify the directories that should be recursively scanned:
+
+```yaml
+user: docker
+group: docker
+paths:
+  - /home/docker
+  - /tmp
+```
+
+`fixuid` will only recurse into a directory as long as it is on the same initial device specified in `paths` and will not recurse into directories mounted on other devices.  This includes Docker volumes.  If you want `fixuid` to run on the root Docker filesystem and a Docker volume at `/home/docker/.cache`, your configuration should include:
+
+```yaml
+user: docker
+group: docker
+paths:
+  - /
+  - /home/docker/.cache
+```
+
 ## Run in Startup Script instead of Entrypoint
 
 You can run `fixuid` as part of your container's startup script.  `fixuid` will `export HOME=/path/to/home` if $HOME is the default value of `/`, so be sure to evaluate the output of `fixuid` when running as a script.

--- a/docker/.gitignore
+++ b/docker/.gitignore
@@ -1,3 +1,2 @@
 /fs-stage/usr/local/bin/fixuid
 stage
-

--- a/docker/fs-stage/tmp/space dir/space file
+++ b/docker/fs-stage/tmp/space dir/space file
@@ -1,0 +1,1 @@
+space file

--- a/docker/fs-stage/tmp/space file
+++ b/docker/fs-stage/tmp/space file
@@ -1,0 +1,1 @@
+space file

--- a/docker/fs-stage/usr/local/bin/fixuid-test.sh
+++ b/docker/fs-stage/usr/local/bin/fixuid-test.sh
@@ -27,7 +27,9 @@ then
     rc=1
 fi
 
-files="/tmp/test-dir /tmp/test-dir/test-file /tmp/test-file /home/docker /home/docker/aaa /home/docker/zzz"
+OLD_IFS="$IFS"
+IFS="|"
+files="/tmp/test-dir|/tmp/test-dir/test-file|/tmp/test-file|/home/docker|/home/docker/aaa|/home/docker/zzz|/tmp/space dir|/tmp/space dir/space file|/tmp/space file"
 for file in $files
 do
     file_user=$(stat -c "%U" $file)
@@ -44,6 +46,7 @@ do
         rc=1
     fi
 done
+IFS="$OLD_IFS"
 
 if [ "$user" = "root" ]
 then

--- a/docker/fs-stage/usr/local/bin/fixuid-test.sh
+++ b/docker/fs-stage/usr/local/bin/fixuid-test.sh
@@ -6,8 +6,9 @@ expected_group=$2
 if [ ! -f /var/run/fixuid.ran ]
 then
     set -e
-    eval $( fixuid )
+    set_home=$( fixuid )
     set +e
+    eval $set_home
 fi
 
 rc=0

--- a/fixuid.go
+++ b/fixuid.go
@@ -179,7 +179,14 @@ func main() {
 	// search entire filesystem and chown containerUID:containerGID to runtimeUID:runtimeGID
 	if needChown {
 
-		var pathDev uint64
+		// proccess /proc/mounts
+		mounts, err := parseProcMounts()
+		if err != nil {
+			logger.Fatalln(err)
+		}
+
+		// store the current mountpoint
+		var mountpoint string
 
 		// this function is called for every file visited
 		visit := func(filePath string, fileInfo os.FileInfo, err error) error {
@@ -192,15 +199,15 @@ func main() {
 				return nil
 			}
 
-			// stat file to determine UID, GID, and device
+			// stat file to determine UID and GID
 			sys, ok := fileInfo.Sys().(*syscall.Stat_t)
 			if !ok {
 				logger.Println("cannot stat " + filePath)
 				return filepath.SkipDir
 			}
 
-			// prevent recursing into mounts - skip if it is not the same device as /
-			if sys.Dev != pathDev {
+			// prevent recursing into mounts
+			if findMountpoint(filePath, mounts) != mountpoint {
 				if sys.Uid == containerUIDUint32 && sys.Gid == containerGIDUint32 {
 					logger.Println("skipping mounted path " + filePath)
 				}
@@ -224,19 +231,14 @@ func main() {
 		}
 
 		for _, path := range paths {
-			// stat the path to determine device
-			rootInfo, err := os.Stat(path)
+			// stat the path to ensure it exists
+			_, err := os.Stat(path)
 			if err != nil {
 				logger.Println("error accessing path: " + path)
 				logger.Println(err)
 				continue
 			}
-			sys, ok := rootInfo.Sys().(*syscall.Stat_t)
-			if !ok {
-				logger.Println("cannot stat " + path)
-				continue
-			}
-			pathDev = sys.Dev
+			mountpoint = findMountpoint(path, mounts)
 
 			logger.Println("recursively searching path " + path)
 			filepath.Walk(path, visit)
@@ -390,4 +392,38 @@ func updateEtcGroup(group string, oldGID string, newGID string) error {
 	}
 
 	return nil
+}
+
+func parseProcMounts() (map[string]bool, error) {
+	// device mountpoint type options dump fsck
+	// spaces appear as \040
+	file, err := os.Open("/proc/mounts")
+	if err != nil {
+		return nil, err
+	}
+
+	mounts := make(map[string]bool)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		cols := strings.Fields(scanner.Text())
+		if len(cols) >= 2 {
+			mounts[filepath.Clean(strings.Replace(cols[1], "\\040", " ", -1))] = true
+		}
+	}
+	file.Close()
+
+	return mounts, nil
+}
+
+func findMountpoint(path string, mounts map[string]bool) string {
+	path = filepath.Clean(path)
+	var lastPath string
+	for path != lastPath {
+		if _, ok := mounts[path]; ok {
+			return path
+		}
+		lastPath = path
+		path = filepath.Dir(path)
+	}
+	return "/"
 }

--- a/fixuid.go
+++ b/fixuid.go
@@ -102,6 +102,18 @@ func main() {
 	}
 	containerGIDUint32 := uint32(containerGIDInt)
 
+	// validate the paths from the config
+	var paths []string
+	err = rootConfig.Configure(&paths, "paths")
+	if err != nil {
+		switch err.(type) {
+		case *config.ConfigPathError:
+			paths = append(paths, "/")
+		default:
+			logger.Fatalln("key 'paths' is malformed; should be an array of strings in configuration file " + filePath)
+		}
+	}
+
 	// declare uid/gid vars and
 	var oldUID, newUID, oldGID, newGID string
 	needChown := false
@@ -167,29 +179,28 @@ func main() {
 	// search entire filesystem and chown containerUID:containerGID to runtimeUID:runtimeGID
 	if needChown {
 
-		// stat / to determine device
-		rootInfo, err := os.Stat("/")
-		if err != nil {
-			logger.Fatal(err)
-		}
-		sys, ok := rootInfo.Sys().(*syscall.Stat_t)
-		if !ok {
-			logger.Fatal("Cannot stat /")
-		}
-		rootDev := sys.Dev
+		var pathDev uint64
 
 		// this function is called for every file visited
 		visit := func(filePath string, fileInfo os.FileInfo, err error) error {
 
+			// an error to lstat or filepath.readDirNames
+			// see https://github.com/boxboat/fixuid/issues/4
+			if err != nil {
+				logger.Println("error when visiting " + filePath)
+				logger.Println(err)
+				return nil
+			}
+
 			// stat file to determine UID, GID, and device
 			sys, ok := fileInfo.Sys().(*syscall.Stat_t)
 			if !ok {
-				logger.Println("Cannot stat " + filePath)
+				logger.Println("cannot stat " + filePath)
 				return filepath.SkipDir
 			}
 
 			// prevent recursing into mounts - skip if it is not the same device as /
-			if sys.Dev != rootDev {
+			if sys.Dev != pathDev {
 				if sys.Uid == containerUIDUint32 && sys.Gid == containerGIDUint32 {
 					logger.Println("skipping mounted path " + filePath)
 				}
@@ -202,17 +213,35 @@ func main() {
 			// only chown if file is containerUID:containerGID
 			if sys.Uid == containerUIDUint32 && sys.Gid == containerGIDUint32 {
 				logger.Println("chown " + filePath)
-				chownError := syscall.Chown(filePath, runtimeUIDInt, runtimeGIDInt)
-				if chownError != nil {
+				err := syscall.Chown(filePath, runtimeUIDInt, runtimeGIDInt)
+				if err != nil {
 					logger.Println("error changing owner of " + filePath)
 					logger.Println(err)
 				}
-				return chownError
+				return nil
 			}
 			return nil
 		}
 
-		filepath.Walk("/", visit)
+		for _, path := range paths {
+			// stat the path to determine device
+			rootInfo, err := os.Stat(path)
+			if err != nil {
+				logger.Println("error accessing path: " + path)
+				logger.Println(err)
+				continue
+			}
+			sys, ok := rootInfo.Sys().(*syscall.Stat_t)
+			if !ok {
+				logger.Println("cannot stat " + path)
+				continue
+			}
+			pathDev = sys.Dev
+
+			logger.Println("recursively searching path " + path)
+			filepath.Walk(path, visit)
+		}
+
 	}
 
 	// mark the script as ran

--- a/test.sh
+++ b/test.sh
@@ -98,20 +98,20 @@ echo "\ndebian should not chown mount"
 docker run --rm -v $(pwd)/docker/fs-stage/tmp:/home/docker/mnt-dir -v $(pwd)/docker/fs-stage/tmp/test-file:/home/docker/mnt-file -u 1234:1234 fixuid-debian sh -c "fixuid-test.sh docker docker && fixuid-mount-test.sh $(id -u) $(id -g)"
 
 
-printf "\npaths:\n  - /\n  - /home/docker\n  - /does/not/exist" >> docker/alpine/stage/etc/fixuid/config.yml
-printf "\npaths:\n  - /\n  - /home/docker\n  - /does/not/exist" >> docker/centos/stage/etc/fixuid/config.yml
-printf "\npaths:\n  - /\n  - /home/docker\n  - /does/not/exist" >> docker/debian/stage/etc/fixuid/config.yml
+printf "\npaths:\n  - /\n  - /home/docker\n  - /tmp/space dir\n  - /does/not/exist" >> docker/alpine/stage/etc/fixuid/config.yml
+printf "\npaths:\n  - /\n  - /home/docker\n  - /tmp/space dir\n  - /does/not/exist" >> docker/centos/stage/etc/fixuid/config.yml
+printf "\npaths:\n  - /\n  - /home/docker\n  - /tmp/space dir\n  - /does/not/exist" >> docker/debian/stage/etc/fixuid/config.yml
 docker-compose build
 
 echo "\nalpine 1001:1001 cmd"
-docker run --rm -u 1001:1001 -v /home/docker fixuid-alpine fixuid-test.sh docker docker
+docker run --rm -u 1001:1001 -v /home/docker -v "/tmp/space dir" fixuid-alpine fixuid-test.sh docker docker
 echo "\ncentos 1001:1001 cmd"
-docker run --rm -u 1001:1001 -v /home/docker fixuid-centos fixuid-test.sh docker docker
+docker run --rm -u 1001:1001 -v /home/docker -v "/tmp/space dir" fixuid-centos fixuid-test.sh docker docker
 echo "\ndebian 1001:1001 cmd"
-docker run --rm -u 1001:1001 -v /home/docker fixuid-debian fixuid-test.sh docker docker
+docker run --rm -u 1001:1001 -v /home/docker -v "/tmp/space dir" fixuid-debian fixuid-test.sh docker docker
 echo "\nalpine 1001:1001 entrypoint"
-docker run --rm -u 1001:1001 -v /home/docker --entrypoint fixuid fixuid-alpine fixuid-test.sh docker docker
+docker run --rm -u 1001:1001 -v /home/docker -v "/tmp/space dir" --entrypoint fixuid fixuid-alpine fixuid-test.sh docker docker
 echo "\ncentos 1001:1001 entrypoint"
-docker run --rm -u 1001:1001 -v /home/docker --entrypoint fixuid fixuid-centos fixuid-test.sh docker docker
+docker run --rm -u 1001:1001 -v /home/docker -v "/tmp/space dir" --entrypoint fixuid fixuid-centos fixuid-test.sh docker docker
 echo "\ndebian 1001:1001 entrypoint"
-docker run --rm -u 1001:1001 -v /home/docker --entrypoint fixuid fixuid-debian fixuid-test.sh docker docker
+docker run --rm -u 1001:1001 -v /home/docker -v "/tmp/space dir" --entrypoint fixuid fixuid-debian fixuid-test.sh docker docker

--- a/test.sh
+++ b/test.sh
@@ -96,3 +96,22 @@ echo "\ncentos should not chown mount"
 docker run --rm -v $(pwd)/docker/fs-stage/tmp:/home/docker/mnt-dir -v $(pwd)/docker/fs-stage/tmp/test-file:/home/docker/mnt-file -u 1234:1234 fixuid-centos sh -c "fixuid-test.sh docker docker && fixuid-mount-test.sh $(id -u) $(id -g)"
 echo "\ndebian should not chown mount"
 docker run --rm -v $(pwd)/docker/fs-stage/tmp:/home/docker/mnt-dir -v $(pwd)/docker/fs-stage/tmp/test-file:/home/docker/mnt-file -u 1234:1234 fixuid-debian sh -c "fixuid-test.sh docker docker && fixuid-mount-test.sh $(id -u) $(id -g)"
+
+
+printf "\npaths:\n  - /\n  - /home/docker\n  - /does/not/exist" >> docker/alpine/stage/etc/fixuid/config.yml
+printf "\npaths:\n  - /\n  - /home/docker\n  - /does/not/exist" >> docker/centos/stage/etc/fixuid/config.yml
+printf "\npaths:\n  - /\n  - /home/docker\n  - /does/not/exist" >> docker/debian/stage/etc/fixuid/config.yml
+docker-compose build
+
+echo "\nalpine 1001:1001 cmd"
+docker run --rm -u 1001:1001 -v /home/docker fixuid-alpine fixuid-test.sh docker docker
+echo "\ncentos 1001:1001 cmd"
+docker run --rm -u 1001:1001 -v /home/docker fixuid-centos fixuid-test.sh docker docker
+echo "\ndebian 1001:1001 cmd"
+docker run --rm -u 1001:1001 -v /home/docker fixuid-debian fixuid-test.sh docker docker
+echo "\nalpine 1001:1001 entrypoint"
+docker run --rm -u 1001:1001 -v /home/docker --entrypoint fixuid fixuid-alpine fixuid-test.sh docker docker
+echo "\ncentos 1001:1001 entrypoint"
+docker run --rm -u 1001:1001 -v /home/docker --entrypoint fixuid fixuid-centos fixuid-test.sh docker docker
+echo "\ndebian 1001:1001 entrypoint"
+docker run --rm -u 1001:1001 -v /home/docker --entrypoint fixuid fixuid-debian fixuid-test.sh docker docker

